agentrs/src/ws/rdp_message_processor.rs

```diff
@@ -0,0 +1,285 @@
+    pub async fn process_messages(
+        self,
+        mut ws_receiver: futures::stream::SplitStream<
+            WebSocketStream<tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>>,
+        >,
+    ) -> anyhow::Result<()> {
+        info!("> Starting to receive messages from gateway...");
+
+        while let Some(msg) = ws_receiver.next().await {
+            match msg? {
+                Message::Binary(data) => {
+                    if let Err(e) = self.handle_binary_message(data.into()).await {
+                        error!("> Error handling binary message: {}", e);
+                    }
+                }
+                Message::Text(text) => {
+                    debug!("> Text from gateway: {}", text);
+                }
+                Message::Close(frame) => {
+                    debug!("> Gateway closed connection: {:?}", frame);
+                    break;
+                }
+                Message::Ping(data) => {
+                    if let Err(e) = self.handle_ping(data.into()).await {
+                        error!("> Failed to respond to ping: {}", e);
+                    }
+                }
+                Message::Pong(_) => {
+                    debug!("> Pong from gateway");
+                }
+                Message::Frame(_) => {
+                    // Handle raw frames if needed
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_binary_message(&self, data: Vec<u8>) -> anyhow::Result<()> {
+        let Some((header, header_len)) = Header::decode(&data) else {
+            info!("> Received data without valid header, ignoring");
+            return Ok(());
+        };
+
+        if header_len > data.len() {
+            error!("> Invalid header length: {} > {}", header_len, data.len());
+            return Ok(());
+        }
+
+        let payload = &data[header_len..];
+
+        // Try to parse as JSON message first
+        if let Ok(message) = serde_json::from_slice::<Value>(payload) {
+            if let Some(message_type) = message.clone().get("message_type").and_then(|v| v.as_str())
+            {
+                return self
+                    .handle_json_message(header, message_type, message)
+                    .await;
+            }
+        }
+
+        // Otherwise treat as RDP data
+        self.handle_rdp_data(header, payload).await
+    }
+
+    async fn handle_json_message(
+        &self,
+        header: Header,
+        message_type: &str,
+        message: Value,
+    ) -> anyhow::Result<()> {
+        match message_type {
+            "session_started" => {
+                info!(
+                    "> Session {} started, processing connection info...",
+                    header.sid
+                );
+                self.handle_session_started(header, message).await
+            }
+            _ => {
+                info!("> Unknown message type: {}", message_type);
+                Ok(())
+            }
+        }
```

Comment:
TODO: improve this abstraction so that it is not RDP specific, so we could add more protocols easily